driver/src/orderbook/streamed/updating_orderbook.rs

```diff
@@ -7,177 +7,144 @@ use crate::{
-    orderbook_ready: Arc<AtomicBool>,
-    // When this struct is dropped this sender will be dropped which makes the updater thread stop.
-    _exit_tx: oneshot::Sender<()>,
+    contract: Arc<dyn StableXContract + Send + Sync>,
+    web3: Web3,
+    /// We need a mutex because otherwise the struct wouldn't be Sync which is needed because we use
+    /// the orderbook in multiple threads. The mutex is locked in `get_auction_data` while the
+    /// orderbook is updated with new events.
+    context: Mutex<Context>,
 }
 
-impl UpdatingOrderbook {
-    pub fn new(contract: Arc<dyn StableXContract + Send + Sync>, web3: Web3) -> Self {
-        let orderbook = Arc::new(Mutex::new(Orderbook::default()));
-        let orderbook_clone = orderbook.clone();
-        let orderbook_ready = Arc::new(AtomicBool::new(false));
-        let orderbook_ready_clone = orderbook_ready.clone();
-        let (exit_tx, exit_rx) = oneshot::channel();
+struct Context {
+    orderbook: Orderbook,
+    last_handled_block: u64,
+    block_timestamp_reader: CachedBlockTimestampReader<Web3>,
+}
 
-        std::thread::spawn(move || {
-            let result = futures::executor::block_on(update_with_events_forever(
-                web3.clone(),
-                contract,
-                orderbook_clone,
-                orderbook_ready_clone,
-                CachedBlockTimestampReader::new(web3),
-                exit_rx,
-            ));
-            if let Err(err) = result {
-                log::error!("event based orderbook failed: {:?}", err);
-                // TODO: implement a retry mechanism
-                // For now we error the program so force a restart of the whole driver because
-                // without a retry we would be stuck with an outdated orderbook forever.
-                // Sleep for one second, so that we have time to flush the logs.
-                thread::sleep(Duration::from_secs(1));
-                process::exit(1);
-            }
-        });
+impl UpdatingOrderbook {
+    /// Blocks on attempting to build the orderbook. After the initial build the orderbook is
+    /// updated every time `get_auction_data` is called.
+    pub fn new(contract: Arc<dyn StableXContract + Send + Sync>, web3: Web3) -> Result<Self> {
+        let result = Self {
+            contract,
+            web3: web3.clone(),
+            context: Mutex::new(Context {
+                orderbook: Orderbook::default(),
+                last_handled_block: 0,
+                block_timestamp_reader: CachedBlockTimestampReader::new(web3),
+            }),
+        };
 
-        Self {
-            orderbook,
-            orderbook_ready,
-            _exit_tx: exit_tx,
+        // Perform the initial update.
+        {
+            let mut context = result
+                .context
+                .lock()
+                .map_err(|err| anyhow!("poison error: {}", err))?;
+            futures::executor::block_on(result.update_with_events(&mut context))?;
```

Comment:
nit: can we use https://github.com/gnosis/dex-services/blob/affcdbb27d20f1b64c0d01dc2d15af235311ec9a/driver/src/util.rs#L44-L46